File: packages/client/src/core/resource.ts
/**
 * Serializable resource references for the managed-agent Resources primitive
 * defined in `docs/proposals/client-primitives.md` and mapped in
 * `docs/explorations/managed-agents-mapping.md`.
 *
 * Mirrors the Rust `ResourceSourceRef` + `PublishedResourceRef` split in
 * `crates/fireline-resources`.
 */
export type HostId = string

export type StreamFsMode = 'snapshotReadOnly' | 'liveReadOnly' | 'liveReadWrite'

export type ResourceSourceRef =
  | {
      readonly kind: 'localPath'
      readonly host_id: HostId
      readonly path: string
    }
  | {
      readonly kind: 's3'
      readonly bucket: string
      readonly key: string
      readonly region: string
      readonly endpoint_url?: string
    }
  | {
      readonly kind: 'gcs'
      readonly bucket: string
      readonly key: string
    }
  | {
      readonly kind: 'dockerVolume'
      readonly host_id: HostId
      readonly volume_name: string
      readonly path_within_volume: string
    }
  | {
      readonly kind: 'durableStreamBlob'
      readonly stream: string
      readonly key: string
    }
  | {
      readonly kind: 'streamFs'
      readonly source_ref: string
      readonly revision?: string
      readonly mode: StreamFsMode
    }
  | {
      readonly kind: 'ociImageLayer'
      readonly image: string
      readonly path: string
    }
  | {
      readonly kind: 'gitRepo'
      readonly url: string
      readonly ref: string
      readonly path: string
    }
  | {
      readonly kind: 'httpUrl'
      readonly url: string
      readonly headers?: Readonly<Record<string, string>>
    }

export type PublishedResourceRef = {
  readonly source_ref: ResourceSourceRef
  readonly mount_path: string
  readonly read_only?: boolean
}

export type ResourceRef = PublishedResourceRef

File: crates/fireline-resources/src/resource.rs
use std::collections::HashMap;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

pub type HostId = String;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum StreamFsMode {
    SnapshotReadOnly,
    LiveReadOnly,
    LiveReadWrite,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ResourceSourceRef {
    LocalPath {
        host_id: HostId,
        path: PathBuf,
    },
    S3 {
        bucket: String,
        key: String,
        region: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        endpoint_url: Option<String>,
    },
    Gcs {
        bucket: String,
        key: String,
    },
    DockerVolume {
        host_id: HostId,
        volume_name: String,
        path_within_volume: PathBuf,
    },
    DurableStreamBlob {
        stream: String,
        key: String,
    },
    StreamFs {
        source_ref: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        revision: Option<String>,
        mode: StreamFsMode,
    },
    OciImageLayer {
        image: String,
        path: PathBuf,
    },
    GitRepo {
        url: String,
        r#ref: String,
        path: PathBuf,
    },
    HttpUrl {
        url: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        headers: Option<HashMap<String, String>>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct PublishedResourceRef {
    pub source_ref: ResourceSourceRef,
    pub mount_path: PathBuf,
    #[serde(default = "default_read_only")]
    pub read_only: bool,
}

pub type ResourceRef = PublishedResourceRef;

const fn default_read_only() -> bool {
    true
}
